lib/value/src/kind/merge.rs

```diff
@@ -0,0 +1,408 @@
+                let last_index = lhs.known().keys().max().map(|i| *i + 1).unwrap_or_default();
+
+                let (rhs_known, rhs_unknown) = rhs.into_parts();
+
+                let mut known = BTreeMap::default();
+                for (index, kind) in rhs_known {
+                    let index = index + last_index;
+                    known.insert(index, kind);
+                }
+
+                lhs.merge(Collection::from_parts(known, rhs_unknown), strategy);
+            }
+
+            (Some(lhs), Some(rhs)) => lhs.merge(rhs, strategy),
+            _ => {}
+        }
+    }
+}
+
+/// The strategy to apply to the merge between two `Kind`s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct Strategy {
+    /// The maximum depth at which to merge nested `Kind`s.
+    ///
+    /// This can be either "shallow" or "deep", meaning don't merge nested collections, or do merge
+    /// them.
+    pub depth: Depth,
+
+    /// The strategy used when merging array indices.
+    ///
+    /// This can either be "keep" or "append", meaning do not update array indices, or append them
+    /// to the end of the `self` array.
+    pub indices: Indices,
+}
+
+/// The choice of "depth" to apply when merging two [`Kind`]s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Depth {
+    /// Do a "shallow" merge for a collection.
+    ///
+    /// Meaning, only the first-level of elements in the collection are merged. If the collection
+    /// contains nested collections, its elements aren't merged, but the collection is swapped out
+    /// for the new one.
+    Shallow,
+
+    /// Do a "deep" merge for a collection.
+    ///
+    /// Meaning, collections are recursively merged.
+    Deep,
+}
+
+impl Depth {
+    /// Check if `shallow` strategy is enabled.
+    #[must_use]
+    pub fn is_shallow(self) -> bool {
+        matches!(self, Self::Shallow)
+    }
+
+    /// Check if `deep` strategy is enabled.
+    #[must_use]
+    pub fn is_deep(self) -> bool {
+        matches!(self, Self::Deep)
+    }
+}
+
+/// The action to take for arrays and their indices when merging two [`Kind`]s.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Indices {
+    /// When merging two arrays, keep their respective indices, potentially merging two `Kinds`
+    /// assigned to the same index.
+    Keep,
+
+    /// When merging two arrays, append the indices of the `other` array after the indices of the
+    /// `self` array.
+    ///
+    /// Meaning, if `self` has index `0` and `other has index `0`, then the index of `other` is
+    /// changed to `1`.
+    Append,
+}
+
+impl Indices {
+    /// Check if `keep` strategy is enabled.
+    #[must_use]
+    pub fn is_keep(self) -> bool {
+        matches!(self, Self::Keep)
+    }
+
+    /// Check if `append` strategy is enabled.
+    #[must_use]
+    pub fn is_append(self) -> bool {
+        matches!(self, Self::Append)
+    }
+}
+
+impl std::ops::BitOr for Kind {
```

Comment:
I think elsewhere in the PR you imported this at the top, is there a reason why they weren't all done the same?

Reply:
No, will check for consistency.